mesa/visualization/mpl_space_drawing.py

```diff
@@ -47,69 +47,31 @@
 def collect_agent_data(
     space: OrthogonalGrid | HexGrid | Network | ContinuousSpace | VoronoiGrid,
     agent_portrayal: Callable,
-    color="tab:blue",
-    size=25,
-    marker="o",
-    zorder: int = 1,
-):
-    """Collect the plotting data for all agents in the space.
-
-    Args:
-        space: The space containing the Agents.
-        agent_portrayal: A callable that is called with the agent and returns a dict
-        color: default color
-        size: default size
-        marker: default marker
-        zorder: default zorder
-
-    agent_portrayal should return a dict, limited to size (size of marker), color (color of marker), zorder (z-order),
-    marker (marker style), alpha, linewidths, and edgecolors
-
-    """
-    arguments = {
-        "s": [],
-        "c": [],
-        "marker": [],
-        "zorder": [],
-        "loc": [],
-        "alpha": [],
-        "edgecolors": [],
-        "linewidths": [],
-    }
+    default_color="tab:blue",
+    default_size=25,
+    default_marker="o",
+    default_zorder: int = 1,
+) -> list[AgentPortrayalStyle]:
+    """Collect agent portrayal data efficiently in one loop."""
+    agent_data_list = []
 
     for agent in space.agents:
-        portray = agent_portrayal(agent)
-        loc = agent.pos
-        if loc is None:
-            loc = agent.cell.coordinate
-
-        arguments["loc"].append(loc)
-        arguments["s"].append(portray.pop("size", size))
-        arguments["c"].append(portray.pop("color", color))
-        arguments["marker"].append(portray.pop("marker", marker))
-        arguments["zorder"].append(portray.pop("zorder", zorder))
-
-        for entry in ["alpha", "edgecolors", "linewidths"]:
-            with contextlib.suppress(KeyError):
-                arguments[entry].append(portray.pop(entry))
-
-        if len(portray) > 0:
-            ignored_fields = list(portray.keys())
-            msg = ", ".join(ignored_fields)
-            warnings.warn(
-                f"the following fields are not used in agent portrayal and thus ignored: {msg}.",
-                stacklevel=2,
+        portrayal_dict = agent_portrayal(agent) or {}
+
+        agent_data_list.append(
+            AgentPortrayalStyle(
```

Comment:
This is wasteful because in the current code you already allways get an `AgentPortrayalStyle` instance.